jmclient/jmclient/blockchaininterface.py

```diff
@@ -1,179 +1,308 @@
-
-import abc
 import ast
+import binascii
 import random
 import sys
 import time
+from abc import ABC, abstractmethod
 from decimal import Decimal
-import binascii
+from typing import Callable, Generator, List, Optional, Union
+
 from twisted.internet import reactor, task
-from jmbase import bintohex, hextobin, stop_reactor
-import jmbitcoin as btc
 
-from jmclient.jsonrpc import JsonRpcConnectionError, JsonRpcError
-from jmclient.configure import jm_single
+import jmbitcoin as btc
+from jmbase import bintohex, hextobin, stop_reactor
 from jmbase.support import get_log, jmprint, EXIT_FAILURE
+from jmclient.jsonrpc import JsonRpc, JsonRpcConnectionError, JsonRpcError
+from jmclient.configure import jm_single
 
 
 # an inaccessible blockheight; consider rewriting in 1900 years
 INF_HEIGHT = 10**8
 
 log = get_log()
 
-class BlockchainInterface(object):
-    __metaclass__ = abc.ABCMeta
+class BlockchainInterface(ABC):
 
-    def __init__(self):
+    def __init__(self) -> None:
         pass
 
-    @abc.abstractmethod
-    def is_address_imported(self, addr):
+    @abstractmethod
+    def is_address_imported(self, addr: str) -> bool:
         """checks that address is already imported"""
 
-    @abc.abstractmethod
-    def is_address_labeled(self, utxo, walletname):
+    @abstractmethod
+    def is_address_labeled(self, utxo: dict, walletname: str) -> bool:
         """checks that UTXO belongs to the JM wallet"""
 
-    @abc.abstractmethod
-    def pushtx(self, txhex):
-        """pushes tx to the network, returns False if failed"""
-
-    @abc.abstractmethod
-    def query_utxo_set(self, txouts, includeconfs=False):
-        """
-        takes a utxo or a list of utxos
-        returns None if they are spend or unconfirmed
-        otherwise returns value in satoshis, address and output script
-        optionally return the coin age in number of blocks
+    @abstractmethod
+    def pushtx(self, txbin: bytes) -> bool:
+        """ Given a binary serialized valid bitcoin transaction,
+        broadcasts it to the network.
         """
-        # address and output script contain the same information btw
 
-    @abc.abstractmethod
-    def estimate_fee_per_kb(self, N):
-        '''Use the blockchain interface to 
-        get an estimate of the transaction fee per kb
-        required for inclusion in the next N blocks.
-	'''
+    @abstractmethod
+    def query_utxo_set(self, txouts: Union[dict, List[dict]],
```

Comment:
I believe the first argument can be a *tuple*, not a dict. i.e. it can be (txid, n). So i think this is wrong?

Reply:
See the implementation in `BitcoinCoreInterface`. I understand the ambiguity, because usually in the codebase we have dicts, and the *keys* are tuples (txid, n).

Reply:
Good catch, it looks you are right!